torch/csrc/jit/passes/peephole_list_idioms.cpp

```diff
@@ -21,17 +22,365 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+                  boolean_value_refinements_[non_throwing_block->outputs().at(
+                      i)];
+            }
+          }
+          continue;
+        }
+
+        for (size_t i = 0; i < if_n.outputs().size(); ++i) {
+          if (!(if_n.outputs().at(i)->type() == BoolType::get())) {
+            continue;
+          }
+          Value* true_v = if_n.thenOutputs().at(i);
+          Value* false_v = if_n.elseOutputs().at(i);
+
+          if (!boolean_value_refinements_.count(true_v) &&
+              !boolean_value_refinements_.count(false_v) &&
+              !constant_as<bool>(true_v) && !constant_as<bool>(false_v)) {
+            continue;
+          }
+
+          // if either block has a constant bool output, e.g. `true` on the
+          // true block, then for the `false` value we can take the false
+          // refinements present on the false block and from the other block
+          // output value bc if the output is false it had to have come from the
+          // false block. if len(a) == 5:
+          //     x = len(b) == 4
+          // else:
+          //     x = False
+          // if x is true, then we know both len(a) == 5 and len(b) == 4
+          //
+          // if neither block has a constant bool value, we just take the
+          // intersection of the refinements from boolean outputs.
+          // if cond:
+          //    x = len(a) == 4 and len(b) == 5
+          // else:
+          //    x = len(a) == 4
+          // here, we know if x is true, then len(a) == 4, but not len(b)
+          // == 5, because that refinement is not present in the true block.
+          // TODO: could also take intersection of refinements present in
+          // both blocks, but it's not a real use case.
+
+          // boolean_value_refinements_[value] is safe to access because
+          // BooleanRefinementMapping has a default constructor
+
+          BooleanRefinementMapping out;
+          if (auto maybe_bool = constant_as<bool>(true_v)) {
+            if (*maybe_bool) {
+              out = BooleanRefinementMapping::FalseRefinements(unionRefinements(
+                  boolean_value_refinements_[false_v].false_refine(),
+                  false_block_refinements));
+            } else {
+              out = BooleanRefinementMapping::TrueRefinements(unionRefinements(
+                  boolean_value_refinements_[false_v].true_refine(),
+                  false_block_refinements));
+            }
+          } else if (auto maybe_bool = constant_as<bool>(false_v)) {
+            if (*maybe_bool) {
+              out = BooleanRefinementMapping::FalseRefinements(unionRefinements(
+                  boolean_value_refinements_[true_v].false_refine(),
+                  true_block_refinements));
+            } else {
+              out = BooleanRefinementMapping::TrueRefinements(unionRefinements(
+                  boolean_value_refinements_[true_v].true_refine(),
+                  true_block_refinements));
```

Comment:
I'm not sure I understand logic of these cases. I tried to walk through some examples and my results disagree with the logic here IIUIC:

```
Case A:
x0 = if len(a) == 5:
     |   x1 = len(b) == 4
     | else:
     |   x2 = False

if x0 is False then (len(b) != 4 OR len(a) != 5) ==> no guarantees can be derived
if x0 is True  then (len(b) == 4 AND len(a) == 5)
x0_refinements = {true: {},
                  false: {b->4, a->5}}

Case B:
x0 = if len(a) == 5:
     |   x1 = len(b) == 4
     | else:
     |   x2 = True

if x0 is False then (len(a) == 5 AND len(b) != 4) ==> len(a) is guaranteed to be 5
if x0 is True  then (len(a) != 5 OR  len(b) == 4) ==> no guarantees can be derived
x0_refinements = {true: {a->5}, false: {}}

Case C:
x0 = if len(a) == 5:
     |   x1 = True
     | else:
     |   x2 = len(b) == 4

if x0 is False then (len(a) != 5 AND len(b) != 4) ==> no guarantees can be derived
if x0 is True  then (len(a) == 5 OR  len(b) == 4) ==> no guarantees can be derived
x0_refinements = {true: {}, false: {}}

Case D:
x0 = if len(a) == 5:
     |   x1 = False
     | else:
     |   x2 = len(b) == 4

if x0 is False then (len(a) == 5 OR len(b) != 4) ==> no guarantees can be derived
if x0 is True  then (len(a) != 5 OR len(b) == 4) ==> no guarantees can be derived
x0_refinements = {true: {}, false: {}}
```

Probably I still don't understand how it works - could you please point me to where I made a mistake in these equations?

Reply:
discussed offline